daemon/qrexec-daemon.c

```diff
@@ -752,87 +782,200 @@ static void sanitize_name(char * untrusted_s_signed, char *extra_allowed_chars)
+    int pid = -1;
     struct sockaddr_un daemon_socket_address = {
         .sun_family = AF_UNIX,
         .sun_path = QREXEC_SOCKET_PATH
     };
 
-    daemon_socket = socket(AF_UNIX, SOCK_STREAM, 0);
+    *user = *target = NULL;
+    int daemon_socket = socket(AF_UNIX, SOCK_STREAM, 0);
     if (daemon_socket < 0) {
          PERROR("socket creation failed");
-         return -1;
+         _exit(126);
     }
 
-    result = connect(daemon_socket, (struct sockaddr *) &daemon_socket_address,
-            sizeof(daemon_socket_address));
-    if (result < 0) {
-         PERROR("connection to socket failed");
-         return -1;
-    }
+    int connect_result = connect(daemon_socket,
+                                 (struct sockaddr *) &daemon_socket_address,
+                                 sizeof(daemon_socket_address));
+    if (connect_result != 0) {
+        PERROR("connection to socket failed");
+        assert(connect_result == -1);
+        if (close(daemon_socket))
+            abort();
+        int fds[2];
+        if (socketpair(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0, fds)) {
+            PERROR("socketpair()");
+            _exit(126);
+        }
+        daemon_socket = fds[0];
 
-    command_size = asprintf(&command, "domain_id=%d\n"
-        "source=%s\n"
-        "intended_target=%s\n"
-        "service_and_arg=%s\n"
-        "process_ident=%s\n\n",
-        remote_domain_id, remote_domain_name, target_domain,
-        service_name, request_id->ident);
-    if (command_size < 0) {
-         PERROR("failed to construct request");
-         return -1;
-    }
+        pid = fork();
+        switch (pid) {
+            case -1:
+                LOG(ERROR, "Could not fork!");
+                _exit(126);
+            case 0:
+                if (close(fds[0]))
+                    abort();
+                if (dup2(fds[1], 0) != 0 || dup2(fds[1], 1) != 1) {
+                    PERROR("dup2()");
+                    _exit(126);
+                }
+                if (close(fds[1]))
+                    abort();
+                char remote_domain_id_str[10];
+                int v = snprintf(remote_domain_id_str, sizeof(remote_domain_id_str), "%d",
+                        remote_domain_id);
+                if (v >= 1 && v < (int)sizeof(remote_domain_id_str)) {
+                    execl(policy_program,
+                            "qrexec-policy-exec",
+                            "--no-exec",
+                            "--",
+                            remote_domain_id_str,
+                            remote_domain_name,
+                            target_domain,
+                            service_name,
+                            request_id->ident,
+                            NULL);
+                    PERROR("execl");
+                } else {
+                    PERROR("snprintf");
+                }
+                _exit(126);
+            default:
+                if (close(fds[1]))
+                    abort();
```

Comment:
If you move the fallback to a separate function and wrap response parsing in a loop (executed at most twice), I think you can rather easily have it also in case of invalid response too (which is IMO useful thing to have).